README.md

```diff
@@ -1,4 +1,223 @@
+    maven {
+        url "https://oss.sonatype.org/content/repositories/snapshots/"
+    }
+}
+
+implementation "com.freeletics.coredux:core:1.0.0-SNAPSHOT"
+```
+
+### Additional artifacts
+
+Following additional artifacts are also available:
+- `com.freeletics.coredux:log-android:1.0.0` - provides `LogSink` implementation, that utilizes `android.util.Log` class to print log messages
+
+## What is CoRedux
+
+CoRedux is a predictable state container, that is using same approach as [Redux](https://redux.js.org/) with
+ability to have additional side effects.
+
+Implementation is based on [Kotlin](https://kotlinlang.org/) [coroutines](https://github.com/Kotlin/kotlinx.coroutines)
+and inspired by both [RxRedux](https://github.com/freeletics/rxredux) library
+and [Coroutines in practice by Roman Elizarov](https://www.youtube.com/watch?v=a3agLJQ6vt8) KotlinConf2018 talk.
+
+### Basic concept
+
+Imagine - we need to develop a calculator app that has on new instance of app start initial value (state) `0`.
+Calculator only allows addition, deduction, multiplication and division operations.
+Operations describe what should happen with the calculator current state and can be represented as following input actions:
+```kotlin
+sealed class CalculatorAction {
+    data class Add(val value: Int) : CalculatorAction()
+    data class Deduct(val value: Int) : CalculatorAction()
+    data class Multiply(val value: Int) : CalculatorAction()
+    data class Divide(val value: Int) : CalculatorAction()
+}
+```
+
+With CoRedux you can create a store that will be a single source of current Calculator state
+and will update it on each new incoming `CalculatorAction`:
+```kotlin
+val store = coroutineScope.createStore<Int, CalculatorAction>(
+    name = "Calculator",
+    initialState = 0,
+    reducer = { currentState, newAction ->
+        when (newAction) {
+            is CalculatorAction.Add -> currentState + newAction.value
+            is CalculatorAction.Deduct -> currentState - newAction.value
+            is CalculatorAction.Multiply -> currentState * newAction.value
+            is CalculatorAction.Divide -> currentState / newAction.value
+        }
+    }
+)
+```
+
+Where `reducer` is a special function responsible for managing state. A `reducer` specify
+how the state changes in response to actions sent to the store.
+Remember that actions only describe what happened, but don't describe how the application's state changes.
+That is the job of the `reducer`.
+
+Finally you have to subscribe `StateReceiver` to the `store` instance to get the state updates over time:
+```kotlin
+store.subscribe { state -> updateUI(state) }
+```
+Where `StateReceiver` is a function, that will receive each new update of the state inside `store`.
+More then one `StateReceiver` can subscribe to the same `store` instance.
+
+With default parameters (`launchMode = CoroutineStart.LAZY`), on first subscription, `store` will emit `initialState` to it.
```

Comment:
> If I understood it correctly the fact that CoroutineStart.LAZY is emitting initalState to state receiver is mostly just because otherwise (i.e. with CoroutineStart.DEFAULT) the state initial state will be emitted to an empty list of StateReceiver because store.subscribe() didn't happen fast enough.

Yes, in `CoroutineStart.DEFAULT` case, reducer coroutine will immediately start, set current state to `initialState` and will wait for a new actions. No `StateReceiver`s will receive `initialState`, as no is subscribed at that moment.

> Also, could it be possible that even with CoroutineStart.DEFAULT the initial state could be emitted to StateReceiver i.e. if somehow the reducer is having a delay(10_000) or if initialState = 0.also { delay(10_000) } and store.subscriber() is called on a different coroutinescope or thread?

Using `delay(...)` in `reducer` function or `State.also { .. }` is not possible (needs to be in coroutine context/`suspend fun`). I can think of using, possibly instead, `Thread.sleep(...)`, but I even wonder what will happen. I'd say it is edge case.

> I would advocate for immediately pass the emit the current state no matter if you are the first subscriber or if you subscribed much much later after some actions have already triggered some actions.

I would not change current approach:
- with `CoroutineStart.DEFAULT` - you can use "headless" store and only receive state updates by subscribing/unsubscribing to it. Kind of event bus for particular state.
- if user needs "headless" store, but also wants to receive last state - they can use combination of `CoroutineStart.LAZY` with custom wrapper around `Store` interface (possibly worth to include in core artefact), that will store last state and propagate it to new subscribers.

I would try to better formulate this moment.